aten/src/ATen/native/Convolution.cpp

```diff
@@ -301,6 +301,10 @@ at::Tensor conv1d(
 at::Tensor conv2d(
     const Tensor& input, const Tensor& weight, const Tensor& bias,
     IntArrayRef stride, IntArrayRef padding, IntArrayRef dilation, int64_t groups) {
+  if (input.type_id() == MkldnnCPUTensorId()) {
+    return at::mkldnn_convolution(
+        input, weight, bias, padding, stride, dilation, groups);
+  }
```

Comment:
It's kind of weird this only applies to conv2d.

Reply:
to pile on to @ezyang's point here -- this is also skipping all the useful error checking and utility code that already exists for convolution.  Consider the difference in the experience here:
```
>>> F.conv2d(torch.randn(1,8,4,3,3), torch.randn(1,4,5,5), padding=1)
RuntimeError: Expected 4-dimensional input for 4-dimensional weight [1, 4, 5, 5], but got 5-dimensional input of size [1, 8, 4, 3, 3] instead
```

```
>>> F.conv2d(torch.randn(1,8,4,3,3).to_mkldnn(), torch.randn(1,4,5,5).to_mkldnn(), padding=1)
RuntimeError: std::exception
```

there's already logic in here for when to use MKLDNN -- why can't we just modify that to take into account if we have an MKLDNNTensor?

Also, I believe if you do it this way you shouldn't have to reproduce `expand_param_if_needed` and possibly not even `conv_output_size`.

Reply:
1. mkldnn is not supposed to handle all convNd, that's why here I only added it for conv2d
2. there are a lot of unnecessary things done before going into the use_mkldnn branch and I want to avoid that for mkldnn (e.g. input.contiguous). To be fair the dispatch should really be put in the native_functions.yaml and let codegen handles it, but because of varies kind of autograd reason we now need to put it into the op itself, so it's better to put this dispatch at the very beginning to make it clear (instead of hiding this dispatch somewhere in the already complicated _convolution).
3. you still need conv_output_size, because you need to pass it to ideep

Reply:
Oh also we intentionally to call at::mkldnn_convolution here (instead of directly calling to the ideep_conv2d c++ function) because we need to use the codegen VariableType::mkldnn_convolution to do unwrap.

Reply:
> Oh also we intentionally call at::mkldnn_convolution here (instead of directly calling to the ideep_conv2d c++ function) because we need to use the codegen VariableType::mkldnn_convolution to do unwrap.

Calling `mkldnn_convolution` in this file is correct for the reason that you note.  The question is about when you call it.

As to your points:
> mkldnn is not supposed to handle all convNd, that's why here I only added it for conv2d

Sure, that's fine, you don't need to call it in all cases.  The point is to just move the dispatch logic to the place that already exists so you get error checking and such for free.  None of the existing code unconditionally calls other kernel -- it checks the parameters and dispatches to the 'best' kernel or errors out if it can't be supported.

> there are a lot of unnecessary things done before going into the use_mkldnn branch and I want to avoid that for mkldnn (e.g. input.contiguous).

It sounds like we should fix those because they are wrong?

> To be fair the dispatch should really be put in the native_functions.yaml and let codegen handles it, but because of varies kind of autograd reason we now need to put it into the op itself, so it's better to put this dispatch at the very beginning to make it clear (instead of hiding this dispatch somewhere in the already complicated _convolution).

This is just moving the problem.  You would still need to reproduce all the error checking / setup code.